### PR TITLE
Detect unused structs which derived Default

### LINKED ISSUE #98871: derive(Default) suppresses warn(dead_code)
Given the following code: [[playground]](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cccc52bdc184fe0561a4757b67821202)

```rust
#[derive(Default)]
struct T {}

struct U {}
```

The current output is:

```
warning: struct `U` is never constructed
 --> src/lib.rs:4:8
  |
4 | struct U {}
  |        ^
  |
  = note: `#[warn(dead_code)]` on by default
```

Ideally the output should look like:

```
warning: struct `T` is never constructed
 --> src/lib.rs:2:8
  |
2 | struct T {}
  |        ^
  |
  = note: `#[warn(dead_code)]` on by default

warning: struct `U` is never constructed
 --> src/lib.rs:4:8
  |
4 | struct U {}
  |        ^
```

After the derive, the code looks roughly like

```rust
struct T {}

#[automatically_derived]
impl Default for T {
    #[inline]
    fn default() -> T { T {} }
}

struct U {}
```

As the default implementation is both 1) itself unused and 2) `#[automatically_derived]`, ideally it should not count as a use for suppressing the `dead_code` lint on `T`.

The derived implementations for `Clone`, `Debug`, `PartialEq`, and `Hash` also show this behavior.

I seem to recall a previous change to `#[automatically_derived]` to change the interaction between the unused code lint and derived implementations, but could not find it offhand.

### PR DESCRIPTION
**Pull request by mu001999**
Fixes #98871

### CODE REVIEW

**Comment by workingjubilee** [comment]
This should not flag a variant that is tagged with `#[default]`.

**Comment by workingjubilee** [comment]
odd.

**Comment by workingjubilee** [comment]
oh, weird.

**Comment by workingjubilee** [comment]
oh wait, I see, the attr is on the trait, we bootstrap the library, etc. got it.

**Comment by workingjubilee** [comment]
...That type is constructed: https://github.com/rust-lang/rust/blob/76c73827dcd0b363e60b22c3cef64bde4171bf17/src/tools/clippy/clippy_lints/src/lib.rs#L622

It shouldn't ignore it if `Default::default()` is actually called.

**Comment by workingjubilee** [comment]
Huh. You directly test for this case, so I don't know why the test is passing but the clippy code is failing.

**Comment by workingjubilee** [comment]
oic, publicity.

File: compiler/rustc_passes/src/dead.rs [hunk 1]
```diff
@@ -400,6 +400,31 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
...
+                    // so that don't ignore impl Default for pub Enum and Structs
+                    DefKind::Struct | DefKind::Union if self.tcx.visibility(did).is_public() => {
+                        return false;
+                    }
+                    // don't ignore impl Default for Enums,
+                    // cause we don't know which variant is constructed
+                    DefKind::Enum => return false,
+                    _ => (),
+                };
+            }
```
Comment (michaelwoerister) [inline compiler/rustc_passes/src/dead.rs:426]: So, this block talks about `Default` in the comments but the code is more general. Maybe we can clarify the reasoning here. E.g.
- why do we only check that the method does not have a self parameter, but not anything else about the signature?
- why do we check that the impls self-type is local? Isn't that always the case for automatically derived impls?

**Comment by michaelwoerister** [comment]
Looks like the only use of the default method is behind `#[cfg(panic = "unwind")]` but the target the test is compiled for does not support unwinding. Allowing the lint for the test should be fine.

**Comment by Kobzol** [comment]
This caused a [regression](https://github.com/rust-lang/rust/pull/126965#issuecomment-2191044805) on the `derive` benchmark, which seems expected. I don't understand why this made the compiler 2 MiB (1.6%) bigger though, that seems a bit suspicious.

**Comment by michaelwoerister** [comment]
Regarding the performance regression: we could do the `self.tcx.has_attr(trait_of, sym::rustc_trivial_field_reads)` check before the new more expensive checking. That might make a difference.

**Pull request merged**